js/tweek-client/src/utils.ts

```diff
@@ -37,39 +38,37 @@ export const createFetchClient = ({
   clientSecret,
   requestTimeoutInMillis = 8000,
   onError,
-}: FetchClientConfig) => {
-  const fetchClient = (input: RequestInfo, init: RequestInit = {}) => {
-    let headersPromise: Promise<Record<string, string>>;
+}: FetchClientConfig): ((input: RequestInfo, init?: RequestInit | undefined) => Promise<Response>) => {
+  const fetchClient = async (input: RequestInfo, init: RequestInit = {}) => {
+    let authHeaders: Record<string, string> | undefined;
 
     if (getAuthenticationToken) {
-      headersPromise = Promise.resolve(getAuthenticationToken()).then(t => ({ Authorization: `Bearer ${t}` }));
+      const token = await getAuthenticationToken();
+      authHeaders = { Authorization: `Bearer ${token}` };
     } else if (clientId && clientSecret) {
-      headersPromise = Promise.resolve({ 'X-Client-Id': clientId, 'X-Client-Secret': clientSecret });
-    } else {
-      headersPromise = Promise.resolve({});
+      authHeaders = { 'X-Client-Id': clientId, 'X-Client-Secret': clientSecret };
     }
 
-    let fetchPromise = headersPromise.then(authHeaders =>
-      fetch(input, {
+    try {
+      const response = await fetch(input, {
         ...init,
         headers: {
           ...init.headers,
           ...authHeaders,
         },
-      }),
-    );
+      });
 
-    if (onError) {
-      fetchPromise = fetchPromise.then(response => {
-        if (!response.ok) {
-          setImmediate(() => onError(response));
-        }
+      if (onError && !response.ok) {
+        const error = await parseTweekErrorResponse(response);
+        onError(error);
```

Comment:
We used setImmediate so that it won't block the code, and if the onError throws an error itself it won't stop execution

Reply:
The possibility of `onError` itself throwing actually makes `setImmediate` here very dangerous.
Instead of the error being caught by some layer of the web server it will remain uncaught (because of `setImmediate`) and crash the entire node process.

This can be tested with this code:
```js
const app = require('express')()
const throwErr = () => { throw new Error('err') }
app.get('/safe', (req, res) => throwErr())
app.get('/crash', (req, res) => setImmediate( () => throwErr() ))
app.listen(3000, () => console.log('Listening on port 3000'))
```
Sending a request to `/safe` will return an error but the server will be fine, while a request to `/crash` will crash the node process

Reply:
But onError is per client not per request, it was meant for logging and monitoring, it shouldn't interrupt the normal flow

Reply:
so what about something like this:
```
if (onError && !response.ok) {
        setImmediate(async () => {
            try {
                const error = await parseTweekErrorResponse(response);
                onError(error);
            } catch (err) {
                onError(err);
            }
        });
    }
```

Reply:
I removed the async.. do u think we still need try-catch?
it's safer...